src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,207 @@
-        return false;
+        if (size == 0) {
+            Node<T> newNode = new Node<>(null, value, null);
+            head = newNode;
+            tail = newNode;
+            size++;
+            return true;
+        } else {
+            Node<T> currentNode = head;
+            while (true) {
+                if (currentNode.next == null) {
+                    Node<T> newNode = new Node<>(currentNode, value, null);
+                    currentNode.next = newNode;
+                    tail = newNode;
+                    size++;
+                    return true;
+                }
+                currentNode = currentNode.next;
+            }
+        }
     }
-
+    
     @Override
     public void add(T value, int index) {
-
+        checkIndex(index);
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+            size++;
+        } else {
+            Node<T> currentNode = getNode(index);
+            Node<T> prevNode = currentNode.prev;
+            Node<T> newNode = new Node<>(prevNode, value, currentNode);
+            currentNode.prev = newNode;
+            prevNode.next = newNode;
+            size++;
+        }
     }
-
+    
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T element : list) {
+            add(element);
+        }
+        return true;
     }
-
+    
     @Override
     public T get(int index) {
-        return null;
+        checkIncludeIndex(index);
+        return getNode(index).value;
     }
-
+    
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIncludeIndex(index);
+        Node<T> currentNode = getNode(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
-
+    
     @Override
     public T remove(int index) {
+        checkIncludeIndex(index);
+        if (size == 1 && index == 0) {
+            head = null;
+            tail = null;
+            size = 0;
+        }
+        int i = 0;
+        Node<T> currentNode = head;
+        while (currentNode != null) {
+            if (i == index) {
+                Node<T> nextNode = currentNode.next;
+                Node<T> prevNode = currentNode.prev;
+                if (prevNode == null) {
+                    nextNode.prev = null;
+                    head = nextNode;
+                    size--;
+                    return currentNode.value;
+                }
+                if (nextNode == null) {
+                    prevNode.next = null;
+                    tail = prevNode;
+                    size--;
+                    return currentNode.value;
+                }
+                nextNode.prev = prevNode;
+                prevNode.next = nextNode;
+                size--;
+                return currentNode.value;
```

Comment:
that's repeating logic from remove(T value). Here you can use getNode()and then unlink method